apps-rendering/src/server/insertNewsletterIntoItem.ts

```diff
@@ -0,0 +1,264 @@
+import { OptionKind } from '@guardian/types';
+import type { Body, BodyElement, NewsletterSignUp } from 'bodyElement';
+import { ElementKind } from 'bodyElementKind';
+import type { Item } from 'item';
+import { logger } from 'logger';
+import { Result } from 'result';
+import { stringToPillar } from 'themeStyles';
+
+// ----- Constants ----- //
+
+const enum ElementCategory {
+	'ParagraphText' = 'PARAGRAPH',
+	'BoldParagraphText' = 'BOLD_PARAGRAPH',
+	'NonParagraphText' = 'OTHER TEXT ELEMENT',
+	'Heading' = 'HEADING',
+	'NonText' = 'NON TEXT',
+	'WhiteSpace' = 'WHITE SPACE',
+	'Error' = 'ERROR',
+}
+
+const DEBUG = {
+	log: false,
+	logContent: false,
+};
+
+// ----- pure functions ---//
+
+function categoriseElement(
+	result: Result<string, BodyElement>,
+): ElementCategory {
+	if (!result.isOk()) {
+		// using result.isErr does not convince the compiler that result.value exists
+		return ElementCategory.Error;
+	}
+
+	switch (result.value.kind) {
+		case ElementKind.HeadingThree:
+		case ElementKind.HeadingTwo:
+			return ElementCategory.Heading;
+		case ElementKind.Text: {
+			const { doc } = result.value;
+			if (doc.textContent?.trim().length === 0) {
+				return ElementCategory.WhiteSpace;
+			}
+
+			if (doc.nodeName === 'P') {
+				if (doc.nodeType === 1) {
+					const element = doc as Element;
+					// for paragraphs containing only single <strong> child ("bold text"),
+					// the sign up block can be put before them, but not after them.
+					if (
+						element.children.length === 1 &&
+						element.firstElementChild?.tagName === 'STRONG'
+					) {
+						return ElementCategory.BoldParagraphText;
+					}
+				}
+				return ElementCategory.ParagraphText;
+			} else {
+				return ElementCategory.NonParagraphText;
+			}
+		}
+		default:
+			return ElementCategory.NonText;
+	}
+}
+
+/**
+ * Get the target range of places within a list of visible elements
+ * that the sign-up block can be placed in, being the places within
+ * 3 elements from the middle of the article (by element number only;
+ * not taking account of the length/height of each element)
+ *
+ * @param numberOfElements
+ * @returns The minimum and maximum index for the target range
+ */
+function getRange(numberOfElements: number): [number, number] {
+	const middle = Math.floor(numberOfElements / 2);
+	return [Math.max(0, middle - 3), Math.min(numberOfElements, middle + 3)];
+}
+
+/**
+ *  Find the place to insert a insert a sign-up block in to an article body,
+ *  using these rules:
+ *	 - Don't put straight after bold text
+ *	 - Don't put under headings
+ *	 - Must have plain body text above and below
+ *   - Must be within 3 elements from the middle of the article
+ *   - Should not be next to an adSlot (NOT IMPLEMENTED HERE)
+ *   - The best place is the last place meeting the criteria above.
+ *
+ * @param body an Item.Body
+ * @returns A Result with the insertIndex or an error
+ */
+function findInsertIndex(body: Body): Result<string, number> {
```

Comment:
I had an idea for this function which might simplify things a little bit, but I'd be very interested to hear your thoughts.

I really like the idea of `getRange()`, and so I wanted to see if we could take even more advantage of that return value. Given `getRange()` gives us the first and last possible places to put the newsletter signup, I think we could narrow all processing between these two values.

In pseudocode, the algorithm then looks a bit like this:

```
const [firstPossibleIndex, lastPossibleIndex] = getRange(elements);

const getInsertionPoint(min: number, currentIndex: number, bodyElements: BodyElement[]): Result<string, number> {
  if (currentIndex <= min) return Result.err('could not find a place')
  if (isSuitable(currentIndex, bodyElements)) return Result.ok(currentIndex)
  return getInsertionPoint(min, currentIndex - 1, bodyElements)
}
```

where we start at the last possible point of insertion, and return early if it is correct. otherwise, keep stepping backwards until a place is found or nothing is found.

the definition of `isSuitable` is then a bit different, where it just checks to make sure the element _at_ `currentIndex` and the element _before_ `currentIndex` meet the criteria (not after a heading, not before a bold paragraph etc).